lib/compiler/src/beam_ssa_alias.erl

```diff
@@ -0,0 +1,997 @@
+    {StMap, FuncDb};
+aa_fixpoint([], Order, _OldAliasMap, AAS=#aas{alias_map=AliasMap}, Limit) ->
+    ?DP("**** Things have changed, starting next iteration ****~n"),
+    aa_fixpoint(Order, Order, AliasMap, AAS, Limit - 1).
+
+aa_fun(F, #opt_st{ssa=Linear0,anno=_Anno,args=Args}=St,
+       AAS0=#aas{alias_map=AliasMap0}) ->
+    %% Initially assume all formal parameters are unique for a
+    %% non-exported function, if we have call argument info in the
+    %% AAS, we use it. For an exported function, all arguments are
+    %% assumed to be aliased.
+    ArgsStatus = aa_get_call_args_status(Args, F, AAS0),
+    SS0 = foldl(fun({Var, Status}, Acc) ->
+                        aa_new_ssa_var(Var, Status, Acc)
+                end, #{}, ArgsStatus),
+    ?DP("@@ Args: ~p~n", [ArgsStatus]),
+    {Linear1,SS} = aa_blocks(Linear0, SS0, AAS0),
+    ?DP("SS:~n~s~n~n", [aa_format(SS)]),
+    AAS = aa_merge_call_args_status(SS, AAS0),
+
+    AliasMap = AliasMap0#{ F => SS },
+    {St#opt_st{ssa=Linear1}, AAS#aas{alias_map=AliasMap}}.
+
+%% Main entry point for the alias analysis
+aa_blocks([{L,#b_blk{is=Is0,last=T0}=Blk}|Bs0], SS0, AAS) ->
+    {Is,SS1} = aa_is(Is0, SS0, [], AAS),
+    {T,SS2} = aa_terminator(T0, SS1),
+    {Bs,SS} = aa_blocks(Bs0, SS2, AAS),
+    {[{L,Blk#b_blk{is=Is,last=T}}|Bs],SS};
+aa_blocks([], SS, _) ->
+    {[],SS}.
+
+aa_is([I=#b_set{dst=Dst,op=Op,args=Args,anno=Anno0}|Is], SS0, Acc, AAS) ->
+    SS1 = aa_new_ssa_var(Dst, unique, SS0),
+    SS = case Op of
+             {bif,Bif} ->
+                 aa_bif(Dst, Bif, Args, SS1, AAS);
+             {float,_} ->
+                 SS1;
+             {succeeded,_} ->
+                 SS1;
+             bs_create_bin ->
+                 case Args of
+                     [#b_literal{val=Flag},_,Arg|_] when
+                           Flag =:= private_append ; Flag =:= append ->
+                         case aa_all_dies([Arg], Dst, AAS) of
+                             true ->
+                                 %% Inherit the status of the argument
+                                 aa_join(Dst, Arg, SS1);
+                             false ->
+                                 %% We alias with the surviving arg
+                                 aa_set_aliased([Dst|Args], SS1)
+                         end;
+                     _ ->
+                         %% XXXX Too conservative?
+                         aa_set_aliased([Dst|Args], SS1)
+                 end;
+             bs_extract ->
+                 aa_set_aliased([Dst|Args], SS1);
+             bs_get_tail ->
+                 aa_set_aliased([Dst|Args], SS1);
+             bs_init_writable ->
+                 SS1;
+             bs_match ->
+                 aa_set_aliased([Dst|Args], SS1);
+             bs_start_match ->
+                 [_,Bin] = Args,
+                 aa_set_aliased([Dst,Bin], SS1);
+             bs_test_tail ->
+                 SS1;
+             build_stacktrace ->
+                 SS1; % XXXX
```

Comment:
It means i should have done a `git grep XXXX` before pushing :) I use at as a marker for things I should take a second look at. 

At this particular location my thinking is that `build_stacktrace` is special in that it is only executed when we're exiting and additionally its only argument isn't a term which can contain a binary. I could do a `aa_set_aliased([Dst|Args], SS1)` but I think that is redundant. Correct? I added a comment about that.

Reply:
Any verdict on `build_stacktrace` @bjorng ? The updated comment is in 0af6c67fc3dc4c5072bc77575e4c2cac7fea48b8.

Reply:
`build_stacktrace` is not only used when exiting. The code that follows `build_stacktrace` can do anything, for example print the stack trace to a log file and continue executing.

The argument can potentially contain binaries:

https://github.com/erlang/otp/blob/a2311845930de6dccba93c58b733dc4a75b1d754/erts/emulator/beam/beam_common.c#L1181-L1189

Reply:
So it's even worse, anything live at this point can be captured and aliased.

Reply:
Sort of: a stack trace (currently) only contains the arguments of the operation that crashed, so we only need to consider the arguments of all operations that can crash into the current exception handler.

I'm not sure that tracking it would be worth it though, so maybe the optimization should just give up when encountering these. I don't think that doing so would cost too much since it's pretty rare to use `try/catch` with captured stack traces in the middle of a hot loop that appends binaries.

Reply:
That's not too bad. At least it can't alias terms which are not known when the analysis encounters `build_stacktrace`. Analyzing exactly which operations can crash into the handler sounds complicated, but just aliasing everything is quite simple.Hopefully users will wrap the loop with the try catch and not have the in the body. For example (with the fix in 63fd18d429824ed11e53451a525ef02e6d7ed289):

```Erlang
stacktrace() ->
    X = ..., % something unique
    try
	ex:code_that_fails(),
	X % still unique
    catch
	_:_:Stacktrace ->
	    io:format("Stacktrace: ~p~n", [Stacktrace]),
	    X % now conservatively aliased
    end.
```
Implementing this correction does not remove any private-append transforms in the set of files compiled by diffable.

Are there any plans to include more in the stack trace, i.e. arguments to functions higher on the stack? That would require the stack-crawler (or what it is called) to deep copy terms.